File: src/backend/go/internal/service/todo_item_service.go
package service

import (
	"encoding/csv"
	"errors"
	"io"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/todo/backend/go/internal/dto"
	"github.com/todo/backend/go/internal/models"
	"github.com/todo/backend/go/internal/repository"
	"gorm.io/gorm"
)

// ErrNotFound is returned when a requested entity does not exist.
// Mirrors a NotFoundException / 404 pattern in ASP.NET Core services.
var ErrNotFound = errors.New("todo item not found")

// csvFieldNames is the CSV header row - matches the Python/PHP/Java implementations exactly.
var csvFieldNames = []string{"id", "title", "description", "is_completed", "created_at", "updated_at"}

// TodoItemService defines the business-logic contract.
// Mirrors ITodoItemService in C#.
type TodoItemService interface {
	GetAll(page, pageSize int) (dto.PaginatedResponse[dto.TodoItemResponse], error)
	GetIncomplete(page, pageSize int) (dto.PaginatedResponse[dto.TodoItemResponse], error)
	GetByID(id uint) (dto.TodoItemResponse, error)
	Create(req dto.CreateTodoItemRequest) (dto.TodoItemResponse, error)
	Update(id uint, req dto.UpdateTodoItemRequest) (dto.TodoItemResponse, error)
	MarkComplete(id uint) (dto.TodoItemResponse, error)
	Delete(id uint) error
	ImportCSV(r io.Reader) (dto.ImportResult, error)
	ExportCSV() (string, error)
}

type todoItemService struct {
	repo repository.TodoItemRepository
}

// NewTodoItemService constructs the service with its repository dependency injected.
func NewTodoItemService(repo repository.TodoItemRepository) TodoItemService {
	return &todoItemService{repo: repo}
}

// ── Mapping ───────────────────────────────────────────────────────────────────

func toResponse(m *models.TodoItem) dto.TodoItemResponse {
	return dto.TodoItemResponse{
		ID:          m.ID,
		Title:       m.Title,
		Description: m.Description,
		IsCompleted: m.IsCompleted,
		CreatedAt:   m.CreatedAt,
		UpdatedAt:   m.UpdatedAt,
	}
}

func toPaginated(result repository.PaginatedResult, page, pageSize int) dto.PaginatedResponse[dto.TodoItemResponse] {
	items := make([]dto.TodoItemResponse, len(result.Items))
	for i := range result.Items {
		items[i] = toResponse(&result.Items[i])
	}
	return dto.PaginatedResponse[dto.TodoItemResponse]{
		Items:      items,
		Total:      result.Total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: int(math.Ceil(float64(result.Total) / float64(pageSize))),
	}
}

func (s *todoItemService) getOrNotFound(id uint) (*models.TodoItem, error) {
	item, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, ErrNotFound
	}
	return item, err
}

// ── Queries ───────────────────────────────────────────────────────────────────

func (s *todoItemService) GetAll(page, pageSize int) (dto.PaginatedResponse[dto.TodoItemResponse], error) {
	result, err := s.repo.FindAll((page-1)*pageSize, pageSize)
	if err != nil {
		return dto.PaginatedResponse[dto.TodoItemResponse]{}, err
	}
	return toPaginated(result, page, pageSize), nil
}

func (s *todoItemService) GetIncomplete(page, pageSize int) (dto.PaginatedResponse[dto.TodoItemResponse], error) {
	result, err := s.repo.FindIncomplete((page-1)*pageSize, pageSize)
	if err != nil {
		return dto.PaginatedResponse[dto.TodoItemResponse]{}, err
	}
	return toPaginated(result, page, pageSize), nil
}

func (s *todoItemService) GetByID(id uint) (dto.TodoItemResponse, error) {
	item, err := s.getOrNotFound(id)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	return toResponse(item), nil
}

// ── Commands ──────────────────────────────────────────────────────────────────

func (s *todoItemService) Create(req dto.CreateTodoItemRequest) (dto.TodoItemResponse, error) {
	item := &models.TodoItem{Title: req.Title, Description: req.Description}
	created, err := s.repo.Create(item)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	return toResponse(created), nil
}

func (s *todoItemService) Update(id uint, req dto.UpdateTodoItemRequest) (dto.TodoItemResponse, error) {
	item, err := s.getOrNotFound(id)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	if req.Title != nil {
		item.Title = *req.Title
	}
	if req.Description != nil {
		item.Description = req.Description
	}
	if req.IsCompleted != nil {
		item.IsCompleted = *req.IsCompleted
	}
	updated, err := s.repo.Update(item)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	return toResponse(updated), nil
}

func (s *todoItemService) MarkComplete(id uint) (dto.TodoItemResponse, error) {
	item, err := s.getOrNotFound(id)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	item.IsCompleted = true
	updated, err := s.repo.Update(item)
	if err != nil {
		return dto.TodoItemResponse{}, err
	}
	return toResponse(updated), nil
}

func (s *todoItemService) Delete(id uint) error {
	item, err := s.getOrNotFound(id)
	if err != nil {
		return err
	}
	return s.repo.Delete(item)
}

// ── CSV import/export ─────────────────────────────────────────────────────────

func parseBool(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "y":
		return true
	default:
		return false
	}
}

// ImportCSV reads a CSV file (header: title, description, is_completed) and creates a todo
// item for every valid row. Mirrors import_csv in the Python/PHP/Java implementations.
func (s *todoItemService) ImportCSV(r io.Reader) (dto.ImportResult, error) {
	reader := csv.NewReader(r)
	reader.FieldsPerRecord = -1

	header, err := reader.Read()
	if err != nil {
		if err == io.EOF {
			return dto.ImportResult{Errors: []dto.ImportRowError{}}, nil
		}
		return dto.ImportResult{}, err
	}
	colIndex := make(map[string]int, len(header))
	for i, name := range header {
		colIndex[strings.ToLower(strings.TrimSpace(name))] = i
	}
	cell := func(row []string, name string) string {
		idx, ok := colIndex[name]
		if !ok || idx >= len(row) {
			return ""
		}
		return row[idx]
	}

	imported := 0
	errs := []dto.ImportRowError{}
	rowNumber := 1 // header is row 1
	for {
		row, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return dto.ImportResult{}, err
		}
		rowNumber++

		title := strings.TrimSpace(cell(row, "title"))
		if title == "" {
			errs = append(errs, dto.ImportRowError{Row: rowNumber, Error: "Title is required."})
			continue
		}

		item := &models.TodoItem{
			Title:       title,
			IsCompleted: parseBool(cell(row, "is_completed")),
		}
		if description := strings.TrimSpace(cell(row, "description")); description != "" {
			item.Description = &description
		}

		if _, err := s.repo.Create(item); err != nil {
			return dto.ImportResult{}, err
		}
		imported++
	}

	return dto.ImportResult{Imported: imported, Failed: len(errs), Errors: errs}, nil
}

// ExportCSV renders every todo item as CSV text. Mirrors export_csv in the
// Python/PHP/Java implementations (same header row and column order).
func (s *todoItemService) ExportCSV() (string, error) {
	items, err := s.repo.FindAllItems()
	if err != nil {
		return "", err
	}

	var buf strings.Builder
	w := csv.NewWriter(&buf)
	if err := w.Write(csvFieldNames); err != nil {
		return "", err
	}
	for _, item := range items {
		description := ""
		if item.Description != nil {
			description = *item.Description
		}
		updatedAt := ""
		if item.UpdatedAt != nil {
			updatedAt = item.UpdatedAt.Format(time.RFC3339Nano)
		}
		record := []string{
			strconv.FormatUint(uint64(item.ID), 10),
			item.Title,
			description,
			strconv.FormatBool(item.IsCompleted),
			item.CreatedAt.Format(time.RFC3339Nano),
			updatedAt,
		}
		if err := w.Write(record); err != nil {
			return "", err
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return "", err
	}
	return buf.String(), nil
}

File: src/backend/python/api/services/todo_item_service.py
import csv
import io
from abc import ABC, abstractmethod

from fastapi import HTTPException, UploadFile, status
from openpyxl import Workbook, load_workbook
from sqlalchemy.orm import Session

from shared.models.todo_item import TodoItem
from api.repositories.todo_item_repository import ITodoItemRepository, TodoItemRepository
from api.schemas.todo_item import (
    CreateTodoItemRequest,
    ImportResult,
    ImportRowError,
    PaginatedResponse,
    TodoItemResponse,
    UpdateTodoItemRequest,
)

# ── CSV import/export helpers ───────────────────────────────────────────────────

_CSV_FIELDNAMES = ["id", "title", "description", "is_completed", "created_at", "updated_at"]
_TRUE_VALUES = {"1", "true", "yes", "y"}


def _parse_bool(value: str | None) -> bool:
    return (value or "").strip().lower() in _TRUE_VALUES


def _parse_bool_cell(value: object) -> bool:
    if isinstance(value, bool):
        return value
    return _parse_bool(str(value) if value is not None else None)


class ITodoItemService(ABC):
    """Service interface - mirrors ITodoService in C#."""

    @abstractmethod
    def get_all(self, page: int, page_size: int) -> PaginatedResponse[TodoItemResponse]: ...

    @abstractmethod
    def get_incomplete(self, page: int, page_size: int) -> PaginatedResponse[TodoItemResponse]: ...

    @abstractmethod
    def get_by_id(self, todo_id: int) -> TodoItemResponse: ...

    @abstractmethod
    def create(self, request: CreateTodoItemRequest, actor_user_id: int | None = None) -> TodoItemResponse: ...

    @abstractmethod
    def update(self, todo_id: int, request: UpdateTodoItemRequest, actor_user_id: int | None = None) -> TodoItemResponse: ...

    @abstractmethod
    def delete(self, todo_id: int) -> None: ...

    @abstractmethod
    def mark_complete(self, todo_id: int, actor_user_id: int | None = None) -> TodoItemResponse: ...

    @abstractmethod
    def import_csv(self, file: UploadFile, actor_user_id: int | None = None) -> ImportResult: ...

    @abstractmethod
    def export_csv(self) -> str: ...

    @abstractmethod
    def import_excel(self, file: UploadFile, actor_user_id: int | None = None) -> ImportResult: ...

    @abstractmethod
    def export_excel(self) -> bytes: ...


class TodoItemService(ITodoItemService):
    """Business-logic layer (analogous to an ASP.NET Core service registered via DI)."""

    def __init__(self, repository: ITodoItemRepository) -> None:
        self._repo = repository

    # ── Helpers ───────────────────────────────────────────────────────────────

    def _get_or_404(self, todo_id: int) -> TodoItem:
        todo = self._repo.get_by_id(todo_id)
        if todo is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Todo item {todo_id} not found.")
        return todo

    @staticmethod
    def _to_paginated(items: list[TodoItem], total: int, page: int, page_size: int) -> PaginatedResponse[TodoItemResponse]:
        return PaginatedResponse(
            items=[TodoItemResponse.model_validate(item) for item in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=-(-total // page_size),  # ceiling division
        )

    # ── Queries ───────────────────────────────────────────────────────────────

    def get_all(self, page: int = 1, page_size: int = 20) -> PaginatedResponse[TodoItemResponse]:
        skip = (page - 1) * page_size
        items, total = self._repo.get_all(skip=skip, limit=page_size)
        return self._to_paginated(items, total, page, page_size)

    def get_incomplete(self, page: int = 1, page_size: int = 20) -> PaginatedResponse[TodoItemResponse]:
        skip = (page - 1) * page_size
        items, total = self._repo.get_incomplete(skip=skip, limit=page_size)
        return self._to_paginated(items, total, page, page_size)

    def get_by_id(self, todo_id: int) -> TodoItemResponse:
        todo = self._get_or_404(todo_id)
        return TodoItemResponse.model_validate(todo)

    # ── Commands ──────────────────────────────────────────────────────────────

    def create(self, request: CreateTodoItemRequest, actor_user_id: int | None = None) -> TodoItemResponse:
        todo = TodoItem(title=request.title, description=request.description, created_by_user_id=actor_user_id)
        created = self._repo.add(todo)
        return TodoItemResponse.model_validate(created)

    def update(self, todo_id: int, request: UpdateTodoItemRequest, actor_user_id: int | None = None) -> TodoItemResponse:
        todo = self._get_or_404(todo_id)
        if request.title is not None:
            todo.title = request.title
        if request.description is not None:
            todo.description = request.description
        if request.is_completed is not None:
            todo.is_completed = request.is_completed
        todo.updated_by_user_id = actor_user_id
        updated = self._repo.update(todo)
        return TodoItemResponse.model_validate(updated)

    def delete(self, todo_id: int) -> None:
        todo = self._get_or_404(todo_id)
        self._repo.delete(todo)

    def mark_complete(self, todo_id: int, actor_user_id: int | None = None) -> TodoItemResponse:
        todo = self._get_or_404(todo_id)
        todo.is_completed = True
        todo.updated_by_user_id = actor_user_id
        updated = self._repo.update(todo)
        return TodoItemResponse.model_validate(updated)

    # ── CSV import/export ─────────────────────────────────────────────────────

    def import_csv(self, file: UploadFile, actor_user_id: int | None = None) -> ImportResult:
        raw = file.file.read()
        text = raw.decode("utf-8-sig")
        reader = csv.DictReader(io.StringIO(text))

        imported = 0
        errors: list[ImportRowError] = []
        for row_number, row in enumerate(reader, start=2):  # header is row 1
            title = (row.get("title") or "").strip()
            if not title:
                errors.append(ImportRowError(row=row_number, error="Title is required."))
                continue

            description = (row.get("description") or "").strip() or None
            is_completed = _parse_bool(row.get("is_completed"))

            todo = TodoItem(title=title, description=description, is_completed=is_completed,
                            created_by_user_id=actor_user_id)
            self._repo.add(todo)
            imported += 1

        return ImportResult(imported=imported, failed=len(errors), errors=errors)

    def export_csv(self) -> str:
        items = self._repo.get_all_items()

        buffer = io.StringIO()
        writer = csv.writer(buffer)
        writer.writerow(_CSV_FIELDNAMES)
        for item in items:
            writer.writerow([
                item.id,
                item.title,
                item.description or "",
                item.is_completed,
                item.created_at.isoformat() if item.created_at else "",
                item.updated_at.isoformat() if item.updated_at else "",
            ])
        return buffer.getvalue()

    # ── Excel import/export ────────────────────────────────────────────────────

    def import_excel(self, file: UploadFile, actor_user_id: int | None = None) -> ImportResult:
        raw = file.file.read()
        workbook = load_workbook(io.BytesIO(raw), read_only=True, data_only=True)
        sheet = workbook.active

        rows = sheet.iter_rows(values_only=True)
        header = [str(cell).strip().lower() if cell is not None else "" for cell in next(rows, ())]
        col_index = {name: idx for idx, name in enumerate(header)}

        def _cell(row: tuple, name: str) -> object:
            idx = col_index.get(name)
            return row[idx] if idx is not None and idx < len(row) else None

        imported = 0
        errors: list[ImportRowError] = []
        for row_number, row in enumerate(rows, start=2):  # header is row 1
            if row is None or all(value is None for value in row):
                continue

            title = str(_cell(row, "title") or "").strip()
            if not title:
                errors.append(ImportRowError(row=row_number, error="Title is required."))
                continue

            description = str(_cell(row, "description") or "").strip() or None
            is_completed = _parse_bool_cell(_cell(row, "is_completed"))

            todo = TodoItem(title=title, description=description, is_completed=is_completed,
                            created_by_user_id=actor_user_id)
            self._repo.add(todo)
            imported += 1

        return ImportResult(imported=imported, failed=len(errors), errors=errors)

    def export_excel(self) -> bytes:
        items = self._repo.get_all_items()

        workbook = Workbook()
        sheet = workbook.active
        sheet.title = "Todo Items"
        sheet.append(_CSV_FIELDNAMES)
        for item in items:
            sheet.append([
                item.id,
                item.title,
                item.description or "",
                item.is_completed,
                item.created_at.isoformat() if item.created_at else "",
                item.updated_at.isoformat() if item.updated_at else "",
            ])

        buffer = io.BytesIO()
        workbook.save(buffer)
        return buffer.getvalue()


# ── Dependency factory (used by FastAPI Depends) ──────────────────────────────

def get_todo_service(db: Session) -> ITodoItemService:
    repository = TodoItemRepository(db)
    return TodoItemService(repository)
